libtcmu.c

```diff
@@ -383,18 +383,192 @@ static void tcmu_cdb_debug_info(const struct tcmulib_cmd *cmd)
                 free(buf);
 }
 
-static void *tcmu_cmdproc_thread(void *arg)
+static void _cleanup_mutex_lock(void *arg)
+{
+        pthread_mutex_t *l = arg;
+        pthread_mutex_unlock(l);
+}
+
+static void *aio_work_queue(void *arg)
 {
         struct tcmu_device *dev = arg;
+        struct tcmu_aio_queue *aio_wq = &dev->work_queue;
+
+        while (1) {
+                int ret;
+                struct tcmu_aio_entry *aio_entry;
+
+                pthread_cleanup_push(_cleanup_mutex_lock, (void *)&aio_wq->aio_lock);
+                pthread_mutex_lock(&aio_wq->aio_lock);
+                {
+                        while (list_empty(&aio_wq->aio_queue)) {
+                                pthread_cond_wait(&aio_wq->aio_cond,
+                                                  &aio_wq->aio_lock);
+                        }
+
+                        aio_entry = list_first_entry(&aio_wq->aio_queue,
+                                                     struct tcmu_aio_entry, entry);
+                        tcmulib_track_aio_request(dev, &aio_entry->entry);
```

Comment:
Maybe just a comment somewhere about the locking order requirements (aio_lock must be taken before track_lock) so new devs don't have to dig in the code.